### PR TITLE
Add documentation on trait objects.

### LINKED ISSUE #21707: Document trait objects in the book
Trait objects seem not to be documented at all in the book, which together with their mandatory lifetime bound annotation makes them almost undiscoverable. Fortunately, there are some nice [blog posts](http://huonw.github.io/blog/2015/01/peeking-inside-trait-objects/) by @huonw, but they are unlikely for a user to stumble upon.

Also, the only occurence of 'trait object' in the reference is following sentence, which is strange:

> Likewise, supertrait methods may also be called on trait objects.

### PR DESCRIPTION
**Pull request by steveklabnik**
I started to write up some docs on this, and then realized I was just repeating http://huonw.github.io/blog/2015/01/peeking-inside-trait-objects/ but worse. @huonw previously said that we can use this content if we wanted, so I made some tweaks and integrated it into the book.

Fixes #21707

### CODE REVIEW

**Comment by huonw** [comment]
:heart:

(It's probably better if I don't review, given I wrote a lot of it.)

**Comment by steveklabnik** [comment]
Hehe, yeah, I thought about it, and decided to let @rust-highfive do its thing. The edits are actually very small.

File: src/doc/trpl/static-and-dynamic-dispatch.md [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+}
+```
+
+This has some upsides: static dispatching of any method calls, allowing for
+inlining and hence usually higher performance. It also has some downsides:
+causing code bloat due to many copies of the same function existing in the
+binary, one for each type.
+
+Furthermore, compilers aren’t perfect and may “optimise” code to become slower.
+For example, functions inlined too eagerly will bloating the instruction cache
```
Comment (alexcrichton) [inline src/doc/trpl/static-and-dynamic-dispatch.md]: will end up bloating?

File: src/doc/trpl/static-and-dynamic-dispatch.md [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+```rust
+# mod foo {
+pub struct TraitObject {
+    pub data: *mut (),
+    pub vtable: *mut (),
+}
+# }
+```
+
+[stdraw]: http://doc.rust-lang.org/nightly/std/raw/struct.TraitObject.html
```
Comment (alexcrichton) [inline src/doc/trpl/static-and-dynamic-dispatch.md]: Could this be a relative link instead of absolute?

**Comment by alexcrichton** [comment]
Nice! A few small nits and otherwise r=me

File: src/doc/trpl/static-and-dynamic-dispatch.md [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+
+Furthermore, compilers aren’t perfect and may “optimise” code to become slower.
+For example, functions inlined too eagerly will bloat the instruction cache
+(cache rules everything around us). This is part of the reason that `#[inline]`
+and `#[inline(always)]` should be used carefully, and one reason why using a
+dynamic dispatch is sometimes more efficient.
+
+However, the common case is that it is more efficient to use static dispatch,
+and one can always have a thin statically-dispatched wrapper function that does
+a dynamic, but not vice versa, meaning static calls are more flexible. The
```
Comment (P1start) [inline src/doc/trpl/static-and-dynamic-dispatch.md]: s/does a dynamic/does a dynamic call/?

**Pull request merged**